src/main/scala/org/allenai/pnp/BsoTrainer.scala

```diff
@@ -0,0 +1,217 @@
+import com.jayantkrish.jklol.training.LogFunction
+import com.jayantkrish.jklol.util.KbestQueue
+
+import edu.cmu.dynet._
+import edu.cmu.dynet.dynet_swig._
+
+/**
+ * Beam search trainer implementing a LaSO-like algorithm.
+ * This trainer implements a margin loss between the
+ * lowest-scoring item on the beam and the highest-scoring
+ * correct execution. Thus, a gradient update is performed
+ * every time all correct executions fall off the beam.
+ */
+class BsoTrainer(val epochs: Int, val beamSize: Int, val maxIters: Int,
+    val model: PpModel, val trainer: Trainer, val log: LogFunction) {
+
+  Preconditions.checkArgument(model.locallyNormalized == false)
+
+  import DynetScalaHelpers._
+
+  def train[A](examples: Seq[PpExample[A]]): Unit = {
+    for (i <- 0 until epochs) {
+      var loss = 0.0
+      log.notifyIterationStart(i)
+      for (example <- examples) {
+        loss +=  doExampleUpdate(example)
+      }
+      
+      trainer.update_epoch()
+
+      log.logStatistic(i, "loss", loss)
+      log.notifyIterationEnd(i)
+    }
+  }
+  
+  private def doExampleUpdate[A](example: PpExample[A]): Double = {
+    val cg = new ComputationGraph
+    var loss = 0.0
+
+    val env = example.env
+    val stateCost = if (example.conditionalExecutionScore == null) {
+      ExecutionScore.zero
+    } else {
+      example.conditionalExecutionScore
+    }
+    val graph = model.getComputationGraph(cg)
+
+    val queue = new BsoPpQueue[A](beamSize, stateCost, graph, log)
+    val finished = new BsoPpQueue[A](beamSize, stateCost, graph, log)
+
+    val startEnv = env.setLog(log)
+    queue.offer(example.unconditional, env, 0.0, null, null, env)
+
+    log.startTimer("bso/beam_search")
+    val beam = new Array[SearchState[A]](beamSize)
+    var numIters = 0
+    val losses = ListBuffer[Expression]()
+    while (queue.queue.size > 0 &&
+        (maxIters < 0 || numIters < maxIters)) {
+      // println(numIters + " " + queue.queue.size)
+
+      // TODO: check this
+      val beamSize = queue.queue.size
+      
+      // Check margin constraint.
+      // The highest-scoring correct execution's score must exceed
+      // that of the lowest-scoring execution on beam by a margin of 1.
+
+      // Note that beam(0) is the lowest-scoring element   
+      // val (beamEx, beamCost) = queue.queue.getItems()(0)
+      log.startTimer("bso/beam_search/get_executions")
+      val worstIncorrectEx = if (queue.incorrectQueue.size > 0) {
+        queue.incorrectQueue.getItems()(0) 
+      } else {
+        null
+      }
+
+      val bestCorrectEx = if (queue.correctQueue.size > 0) {
+        queue.correctQueue.getItems.slice(
+            0, queue.correctQueue.size).maxBy(x => x.logProb)
+      } else {
+        null
+      }
+      log.stopTimer("bso/beam_search/get_executions")
+
+      var nextBeamSize = -1
+      if (numIters != 0 && bestCorrectEx != null && worstIncorrectEx != null &&
+          worstIncorrectEx.logProb + 1 > bestCorrectEx.logProb) {
```

Comment:
Yeah ideally client code will be able to select the value instead of 1, i.e., the margin. I have to figure out the right interface for implementing cost functions, then we can fix this.